### PR TITLE
Creates ReviewTable API to query data from database

### LINKED ISSUE #11: [SUBTASK] Add middleware API for retrieving review data
**What feature is this a subtask of**
#3 

**What will be implemented in this subtask**
A new middleware API that will have a single GET request. When requested, it will use the DAO for the new Reviews table to lookup all User/document/average score entries and output them in an array.
  
**What is the acceptance criteria of this subtask**
When tested in a test file, the API successfully returns a JSON array with entries of the form {name: xxx, docId: xxx, score: xxx}, where the score represents the average score of all reviews from that name/docId.

**What files will be created/modified**
Requires a new resource file (i.e. `ReviewResource.java`) that will define a single GET request API, plus a corresponding new test file (i.e. `TestReviewResource.java`).

**What dependencies does this issue have or create**
Depends on #4. Task #8 depends on this issue.

**What is the estimated effort of this subtask**
Approximately 1-2 days

### LINKED ISSUE #12: [SUBTASK] Add hardcoded data to the Review table
**What feature is this a subtask of**
#3 

**What will be implemented in this subtask**
Backend changes to pre-load the review table with hardcoded data.
  
**What is the acceptance criteria of this subtask**
When tested in a file it should set the Review data table with the corresponding values entered as arguments into the PUT request. Can have assertions in a test file to make sure this occurs. 

**What files will be created/modified**
Modifies file created in subtask #7  `dbupdate-028-0.sql`. 

**What dependencies does this issue have or create**
Depends on #4. Task #8 depends on this issue.

**What is the estimated effort of this subtask**
1 day.

### PR DESCRIPTION
**Pull request by aruiz2**
## Issue
Resolves #11
Resolves #12 

## Description
A new middleware API that will have a single GET request. When requested, it will use the DAO for the new Reviews table to lookup all User/document/average score entries and output them in an array.

## Files changed
Requires a new resource file (i.e. `ReviewResource.java`) that will define a single GET request API, plus a corresponding new test file (i.e. `TestReviewResource.java`).

## Notes for reviewers
N/A

## Checklist
Fill out this checklist before merging the PR:
- [x] Created issue
   - [x] Issue has a relevant labels
   - [x] Issue added to the kanban board with its status as In-Progress
   - [x] Issue added to the relevant milestone
   - [x] Issue mentioned in the first section of this PR description
- [x] Testing
   - [x] Added new tests for the changes (or modified existing tests)
   - [x] Passed most recent round of CI
- [x] Review
   - [x] Received first round of review comments
   - [x] Addressed first round of review comments
   - [x] Received second round of review comments
   - [x] Addressed second round of review comments

### CODE REVIEW

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        Assert.assertEquals("John Doe", entry1.getString(username));
+        Assert.assertEquals("doc123", entry1.getString(docId));
+        Assert.assertEquals(5, entry1.getInt(averageRating));
+
+        JsonObject entry2 = json.getJsonArray("ratings").getJsonObject(0);
+        Assert.assertEquals("Florentino Perez", entry2.getString(username));
+        Assert.assertEquals("doc1234", entry2.getString(docId));
+        Assert.assertEquals(10, entry2.getInt(averageRating));
+        log.info(json.toString());
+        log.info("ALEJANDRO'S TEST DID NOT ERROR!\n\n");
```
Comment (aruiz2) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java]: will delete this

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        JsonObject entry1 = json.getJsonArray("ratings").getJsonObject(1);
+        Assert.assertEquals("John Doe", entry1.getString(username));
+        Assert.assertEquals("doc123", entry1.getString(docId));
+        Assert.assertEquals(5, entry1.getInt(averageRating));
+
+        JsonObject entry2 = json.getJsonArray("ratings").getJsonObject(0);
+        Assert.assertEquals("Florentino Perez", entry2.getString(username));
+        Assert.assertEquals("doc1234", entry2.getString(docId));
+        Assert.assertEquals(10, entry2.getInt(averageRating));
+        log.info(json.toString());
```
Comment (aruiz2) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java]: will delete this

File: docs-core/src/main/resources/db/update/dbupdate-028-0.sql [hunk 3]
```diff
@@ -1,14 +1,4 @@
...
-    REV_DELETEDATE_D date,
-    primary key (REV_REVIEWID_C)
-);
+create memory table T_REVIEW ( REV_USERNAME_C varchar(36), REV_RATING_C integer, REV_REVIEWID_C varchar(36) not null, REV_USERID_C varchar(36), REV_DOCID_C varchar(36), REV_REVIEWERID_C varchar(36), REV_CREATEDATE_D date not null, REV_DELETEDATE_D date, primary key (REV_REVIEWID_C) );
 
-insert into T_REVIEW(REV_USERNAME_C, REV_RATING_C, REV_REVIEWID_C, REV_USERID_C, REV_DOCID_C, REV_REVIEWERID_C, REV_CREATEDATE_D, REV_DELETEDATE_D)
-values ("John Doe", 5, "review123", "jdoe", "doc123", "reviewerperson", 09/28/2022, 09/28/2023);
\ No newline at end of file
+insert into T_REVIEW(REV_USERNAME_C, REV_RATING_C, REV_REVIEWID_C, REV_USERID_C, REV_DOCID_C, REV_REVIEWERID_C, REV_CREATEDATE_D, REV_DELETEDATE_D) values('John Doe', 5, 'review123', 'jdoe', 'doc123', 'reviewerperson', NOW(), NOW());
+insert into T_REVIEW(REV_USERNAME_C, REV_RATI`NG_C, REV_REVIEWID_C, REV_USERID_C, REV_DOCID_C, REV_REVIEWERID_C, REV_CREATEDATE_D, REV_DELETEDATE_D) values('Florentino Perez', 10, 'review1234', 'fperez', 'doc1234', 'reviewerperson', NOW(), NOW());
```
Comment (knassre-cmu) [inline docs-core/src/main/resources/db/update/dbupdate-028-0.sql]: Let's add a few more entries

File: docs-core/src/main/java/com/sismics/docs/core/dao/ReviewDao.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /**
+     * Outputs a list of triplets in the form (Name, DocumentId, Avg Score)
+     * which contains the average review score for all user/document combinations.
+     */
+    public List<ReviewAggDto> getAverageScorePerApplication() {
+        List<ReviewAggDto> result = new ArrayList<ReviewAggDto>();
+        EntityManager em = ThreadLocalContext.get().getEntityManager();
+        Query q = em.createQuery("select r.userName, r.docId, AVG(CAST(r.rating as double)) from Review r group by r.userName, r.docId");
```
Comment (knassre-cmu) [inline docs-core/src/main/java/com/sismics/docs/core/dao/ReviewDao.java]: Ugh, idk why this change was not propagated, but we need to add the order part again.

Try modifying the query to this:
```sql
select r.userName, r.docId, AVG(CAST(r.rating as double)) AS avgRating from Review r group by r.userName, r.docId order by avgRating DESC

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        Assert.assertEquals(2, json.getJsonArray("ratings").size());
+        JsonObject entry1 = json.getJsonArray("ratings").getJsonObject(1);
+        Assert.assertEquals("John Doe", entry1.getString(username));
+        Assert.assertEquals("doc123", entry1.getString(docId));
+        Assert.assertEquals(5, entry1.getInt(averageRating));
+
+        JsonObject entry2 = json.getJsonArray("ratings").getJsonObject(0);
+        Assert.assertEquals("Florentino Perez", entry2.getString(username));
+        Assert.assertEquals("doc1234", entry2.getString(docId));
+        Assert.assertEquals(10, entry2.getInt(averageRating));
```
Comment (knassre-cmu) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java]: After we deal with order, swap these so that we have the one with highest rating first

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    @Test
+    public void testReviewResource() {
+        // Login admin
+        String adminToken = clientUtil.login("admin", "admin", false);
+        String username = "username";
+        String docId = "docId";
+        String averageRating = "avgRating";
+        JsonObject json = target().path("/user_rating").request()
+                .cookie(TokenBasedSecurityFilter.COOKIE_NAME, adminToken)
+                .get(JsonObject.class);
```
Comment (knassre-cmu) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:38]: Before testing content, test the length first

File: docs-core/src/main/java/com/sismics/docs/core/dao/ReviewDao.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package com.sismics.docs.core.dao;
```
Comment (knassre-cmu) [inline docs-core/src/main/java/com/sismics/docs/core/dao/ReviewDao.java]: You need to resolve all the merge conflicts with main. Run `git pull origin main` then manually resolve them on your end.

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tests the ReviewResource.java file
+ *
+ * @author aruiz2
+ */
+
+public class TestReviewResource extends BaseJerseyTest {
+    private static final Logger log = LoggerFactory.getLogger(TestReviewResource.class);
```
Comment (knassre-cmu) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:22]: Since you are testing from the hardcoded files, try running mvn test from the main branch. If these tests are run before the ones I wrote, then that's fine. If they run after, it may cause problems because my tests also add things to the database for testing purposes.

**Review by knassre-cmu — COMMENTED** [review]
Left a few comments. Some of the changes are urgent, but most are just testing stuff.

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java (hunk 8 above)
Comment (aruiz2) [reply docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:22]: Should be good, just made the scores higher than yours

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java (hunk 5 above)
Comment (jding712) [inline docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java]: Double check the types for averageRating (decide between double and int)

File: docs-core/src/main/resources/db/update/dbupdate-028-0.sql (hunk 3 above)
Comment (jding712) [reply docs-core/src/main/resources/db/update/dbupdate-028-0.sql]: Some ideas for entries:
- same users
- same reviewer
- different date times

**Review by jding712 — COMMENTED** [review]
Adding more test cases to fully test out average rating feature would be cool.
Looks good!

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java (hunk 5 above)
Comment (aruiz2) [reply docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java]: it is not letting me use getDouble so for the moment I left as getInt

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java (hunk 6 above)
Comment (aruiz2) [reply docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:38]: Since you add stuff to the database and in the future we might too, should we leave the length check aside?

Comment (knassre-cmu) [reply docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:38]: Maybe for an exact check, but maybe we should at least check that the length is at least X, where X is the number of entries you are testing in the lines below

File: docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java (hunk 8 above)
Comment (knassre-cmu) [reply docs-web/src/test/java/com/sismics/docs/rest/TestReviewResource.java:22]: Lol, that works too

**Review by knassre-cmu — APPROVED** [review]
Unless @jding712 has any pressing comments, this LGTM and we can merge ASAP.

**Review by jding712 — APPROVED** [review]
Looks good to merge!
